cachi2/core/package_managers/pip.py

```diff
@@ -1453,8 +1453,117 @@ def _split_hashes_from_options(cls, options: list[str]) -> tuple[list[str], list
         return hashes, reduced_options
 
 
+def _process_req(
+    req: PipRequirement,
+    requirements_file: PipRequirementsFile,
+    require_hashes: bool,
+    pip_deps_dir: RootedPath,
+    download_info: dict[str, Any],
+    dpi: Optional[DistributionPackageInfo] = None,
+) -> dict[str, Any]:
+    download_info["kind"] = req.kind
+    download_info["requirement_file"] = str(requirements_file.file_path.subpath_from_root)
+    download_info["hash_verified"] = False
+
+    def _hash_verify(path: Path, checksum_info: Iterable[ChecksumInfo]) -> None:
+        try:
+            must_match_any_checksum(path, checksum_info)
+        except PackageRejected:
+            path.unlink()
+            log.warning("Download '%s' was removed from the output directory", path.name)
+
+    if dpi:
+        if dpi.should_verify_checksums():
+            _hash_verify(dpi.path, dpi.checksums_to_verify)
+            download_info["hash_verified"] = True
+
+        if dpi.package_type == "sdist":
+            _check_metadata_in_sdist(dpi.path)
+
+    else:
+        if require_hashes or req.kind == "url":
+            hashes = req.hashes or [req.qualifiers.get("cachito_hash", "")]
+            _hash_verify(download_info["path"], list(map(_to_checksum_info, hashes)))
+            download_info["hash_verified"] = True
+
+    log.info(
+        "-- Successfully processed '%s' in path '%s'",
+        req.download_line,
+        download_info["path"].relative_to(pip_deps_dir.root),
+    )
+
+    return download_info
+
+
+def _process_pypi_req(
+    req: PipRequirement,
+    requirements_file: PipRequirementsFile,
+    require_hashes: bool,
+    pip_deps_dir: RootedPath,
+    allow_binary: bool,
+) -> list[dict[str, Any]]:
+    download_infos: list = []
+
+    def _process_req_wrapper(
+        DPIs: list[DistributionPackageInfo], log_string: str
+    ) -> Optional[dict[str, Any]]:
+        for dpi in DPIs:
+            log.info(log_string, dpi.path.name)
+            info: dict[str, Any] = _process_req(
+                req, requirements_file, require_hashes, pip_deps_dir, dpi.download_info, dpi=dpi
+            )
+            return info
+        return None
+
+    artifacts: list[DistributionPackageInfo] = _process_package_distributions(
+        req, pip_deps_dir, allow_binary
+    )
+    artifacts_to_download: list[DistributionPackageInfo] = [a for a in artifacts if a.path.exists()]
+    files: dict[str, Union[str, PathLike[str]]] = {
+        dpi.url: dpi.path for dpi in artifacts_to_download
+    }
+    asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+    if retval := _process_req_wrapper(artifacts_to_download, "-- Processing pypi download '%s'"):
+        download_infos.append(retval)
+    if retval := _process_req_wrapper(
+        [a for a in artifacts if not a.path.exists()], "-- Checking existing pypi artifact '%s'"
```

Comment:
But then you end up in the situation where you have **no** idea as to what's _actually_ been processed. You just have a big blob of what you *attempted* to process. Yes, the logging is more verbose, but it's also **more informative**.

Digging through the logs of failed OSBS builds (for years) definitely taught me that there's such a thing as too much logging :roll_eyes:, but the work I did on forcing the OSBS integration test logs to "tell a story" about what was being **attempted** vs. what was being **completed** convinced me that the volume of logging isn't a problem - it's **what** and **how** you tell an engineer diagnosing a failure that's actually important (obviously there's a balance to be kept here)

Reply:
Note that I'm not after an incomprehensible log dump where you don't know what is being processed, I'm rather contemplating about the importance to report 2 distinct log messages:

- `"-- Processing pypi download  ... "`
- `"-- Checking existing pypi artifact ... "`

where a simple `"Processing <artifact> name"` might perfectly suffice. If both lists to be processed are dumped somewhere early in the processing then you only need to know which one is being processed at a given time, you don't need to know whether we're downloading it or just checking it, right?

Reply:
OK, I've completely refactored this.

However, be prepared for logs with **110 lines** consisting solely of e.g. "2024-03-20 12:26:04,602 INFO Verifying checksums of tokenizers- ... whl" (or 88 for "setproctitle-", 110 also for "safetensors-", etc.) because *that* log message comes from cachi2/core/checksum.py:65, and I don't plan on touching it.